pennylane/estimator/estimate.py

```diff
@@ -0,0 +1,358 @@
+        config (ResourceConfig | None): A ResourceConfig object which modifies default behaviour in the estimation pipeline.
+
+    Returns:
+        Resources | Callable[..., Resources]: The estimated quantum resources required to execute the circuit.
+
+    Raises:
+        TypeError: could not obtain resources for obj of type :code:`type(obj)`
+
+    **Example**
+
+    The resources of a quantum workflow can be estimated by passing the quantum function describing the
+    workflow directly into this function.
+
+    .. code-block:: python
+
+        import pennylane.estimator as qre
+
+        def my_circuit():
+            for w in range(2):
+                qre.Hadamard(wires=w)
+
+            qre.CNOT(wires=[0,1])
+            qre.RX(wires=0)
+            qre.RY(wires=1)
+
+            qre.QFT(num_wires=3, wires=[0, 1, 2])
+            return
+
+    Note that a python function is passed here, not a :class:`~.QNode`. The resources for this
+    workflow are then obtained by:
+
+    >>> import pennylane.estimator as qre
+    >>> config = qre.ResourceConfig()
+    >>> config.set_single_qubit_rot_precision(1e-4)
+    >>> res = qre.estimate(
+    ...     my_circuit,
+    ...     gate_set = qre.DefaultGateSet,
+    ...     config = config,
+    ... )()
+    ...
+    >>> print(res)
+    --- Resources: ---
+    Total qubits: 3
+    Total gates : 279
+    Qubit breakdown:
+     clean qubits: 0, dirty qubits: 0, algorithmic qubits: 3
+    Gate breakdown:
+     {'Hadamard': 5, 'CNOT': 10, 'T': 264}
+
+    """
+    return _estimate_resources_dispatch(obj, gate_set, zeroed, any_state, tight_budget, config)
+
+
+@singledispatch
+def _estimate_resources_dispatch(
+    obj: ResourceOperator | Callable | Resources | list,
+    gate_set: set[str] | None = None,
+    zeroed: int = 0,
+    any_state: int = 0,
+    tight_budget: bool = False,
+    config: ResourceConfig | None = None,
+) -> Resources | Callable[..., Resources]:
+    """Internal singledispatch function for resource estimation."""
+    raise TypeError(
+        f"Could not obtain resources for obj of type {type(obj)}. obj must be one of Resources, Callable, ResourceOperator, or list"
+    )
+
+
+@_estimate_resources_dispatch.register
+def _resources_from_qfunc(
+    obj: Callable,
+    gate_set: set[str] | None = None,
+    zeroed: int = 0,
+    any_state: int = 0,
+    tight_budget: bool = False,
+    config: ResourceConfig | None = None,
+) -> Callable[..., Resources]:
+    """Get resources from a quantum function which queues operations"""
+
+    if isinstance(obj, QNode):
+        raise NotImplementedError("Support for QNodes has not yet been implemented.")
+
+    @wraps(obj)
+    def wrapper(*args, **kwargs):
+        with AnnotatedQueue() as q:
+            obj(*args, **kwargs)
+
+        wire_manager = WireResourceManager(zeroed, any_state, 0, tight_budget)
+        num_algo_qubits = 0
+        circuit_wires = []
+        for op in q.queue:
```

Comment:
Do we want to do this?
```suggestion
        circuit_wires = []
        operations = QuantumScript.from_queue(q)
        for op in operations:
```

Reply:
No. We don't need any of the `QuantumScript` behaviour here.

Reply:
Ignore me then 😄